src/centrality.rs

```diff
@@ -0,0 +1,217 @@
+    }
+}
+
+fn _accumulate_basic(
+    betweenness: &mut Vec<f64>,
+    node_count: usize,
+    sorted_by_distance: &Vec<NodeIndex>,
+    mut predecessors: HashMap<NodeIndex, Vec<NodeIndex>>,
+    sigma: HashMap<usize, i64>,
+    is: usize,
+) {
+    let mut delta = vec![0.0; node_count];
+    for w in sorted_by_distance {
+        let iw = (*w).index();
+        let coeff = (1.0 + delta[iw]) / (sigma[&iw] as f64);
+        let p_w = predecessors.get_mut(&w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (sigma[&iv] as f64) * coeff;
+        }
+        if iw != is {
+            betweenness[iw] += delta[iw];
+        }
+    }
+}
+
+fn _accumulate_endpoints(
+    betweenness: &mut Vec<f64>,
+    node_count: usize,
+    sorted_by_distance: &Vec<NodeIndex>,
+    mut predecessors: HashMap<NodeIndex, Vec<NodeIndex>>,
+    sigma: HashMap<usize, i64>,
+    is: usize,
+) {
+    betweenness[is] += (sorted_by_distance.len() - 1) as f64;
+    let mut delta = vec![0.0; node_count];
+    for w in sorted_by_distance {
+        let iw = (*w).index();
+        let coeff = (1.0 + delta[iw]) / (sigma[&iw] as f64);
+        let p_w = predecessors.get_mut(&w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (sigma[&iv] as f64) * coeff;
+        }
+        if iw != is {
+            betweenness[iw] += delta[iw] + 1.0;
+        }
+    }
+}
+
+fn shortest_path_for_centrality<G>(
+    graph: G,
+    node_s: &G::NodeId,
+) -> (
+    Vec<NodeIndex>,
+    HashMap<G::NodeId, Vec<G::NodeId>>,
+    HashMap<usize, i64>,
+)
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + GraphBase<NodeId = NodeIndex>, // for get() and get_mut()
+{
+    let mut sorted_by_distance: Vec<NodeIndex> = Vec::new(); // a stack
+    #[allow(non_snake_case)]
+    let mut Q: Vec<NodeIndex> = Vec::new(); // a queue
+    let mut predecessors = HashMap::<G::NodeId, Vec<G::NodeId>>::new();
+    let mut sigma = HashMap::<usize, i64>::new();
+    let mut distance = HashMap::<G::NodeId, i64>::new();
+
+    let i_s = graph.to_index(*node_s);
+    let index_s = NodeIndex::new(i_s);
+
+    for node in graph.node_identifiers() {
+        let i = graph.to_index(node);
+        let index = NodeIndex::new(i);
+        predecessors.insert(index, Vec::new());
+        sigma.insert(i, 0);
+        distance.insert(index, -1);
+    }
+    sigma.insert(index_s.index(), 1);
+    distance.insert(index_s, 0);
+    Q.push(index_s);
+    while !Q.is_empty() {
+        let v = Q.remove(0);
+        sorted_by_distance.push(v);
+        //        let neighbors = graph.neighbors_directed(v, Incoming);
+        for w in graph.neighbors(v) {
+            if *(distance.get(&w).unwrap()) < 0 {
+                Q.push(w);
+                distance.insert(w, *(distance.get(&v).unwrap()) + 1);
+            }
+            if *(distance.get(&w).unwrap()) == *(distance.get(&v).unwrap()) + 1
```

Comment:
Since we use distance[w] and distance[v] more than once do you think it makes sense to store them in a temp so we only need to access once. It likely won't have a performance impact but will definitely clean things up a bit so this line a bit so it'll be something like:

```rust
if distance_w == distance_v + 1 {
```

Reply:
I might guess that the compiler would optimize this. But, it would be more clear to store them in temps.

Reply:
Yeah, the compiler would most likely optimize this for us, sometimes I get too used to working in python (where you would need to do this optimization by hand) :)

Reply:
I now have `if distance[&w] == distance[&v] + 1`, which is ok for verbosity. `distance[&w]` may be mutated. That leaves `distance[&v]` which is read twice. I guess the compiler can see that `distance[&v]` is never written to inside the while loop.

Reply:
This is probably resolved with recent edits